doc/function_plotter.de.man
.\" Process this file with
.\" groff -man -Tutf8 function_plotter.de.man
.\"
.
.\" Macros .TQ .EX .EE taken from groff an-ext.tmac
.\" Copyright (C) 2007, 2009 Free Software Foundation, Inc.
.\" You may freely use, modify and/or distribute this file.
.
.\" Continuation line for .TP header.
.de TQ
.  br
.  ns
.  TP \\$1\" no doublequotes around argument!
..
.
.\" Start example.
.de EX
.  nr mE \\n(.f
.  nf
.  nh
.  ft CW
..
.
.
.\" End example.
.de EE
.  ft \\n(mE
.  fi
.  hy \\n(HY
..
.
.TH "function_plotter.lua" 1 2012-07-24 AKFAvatar
.nh
.
.SH NAME
function_plotter.lua \- Funktions-Plotter f\[:u]r Lua-AKFAvatar
.
.SH SYNTAX
.BI "function_plotter.lua [" Funktion ]
.
.SH BESCHREIBUNG
Funktions-Plotter f\[:u]r Lua-AKFAvatar.
Man kann eine Funktion f\[:u]r die Variable
.I x
angeben, die dann gezeichnet wird.
Als Beispiel k\[:o]nnte man \[Bq](x/5)^3\[lq] eingeben.
.PP
Man kann folgende Operatoren verwenden:
.IP +
Plus (Addition)
.IP -
Minus (Substraktion)
.IP *
Mal (Multiplikation)
.IP /
Geteilt (Division)
.IP ^
Hoch (Zum Beispiel \[Bq]x^3\[lq] f\[:u]r x\[S3])
.IP %
Modulo (Der Rest einer Division)
.PP
Gleitkommazahlen kann man wahlweise mit einem Komma, oder einem Punkt schreiben.
Zum Beispiel \[Bq]0,5\[lq] oder \[Bq]0.5\[lq] sind das selbe.
Tausender-Trennzeichen kann man jedoch nicht verwenden.
Die Zahl \(*p kann als \[Bq]pi\[lq] geschrieben werden.
.PP
Man kann folgende Funktionen verwenden:
.TP 10
.BI sqrt( x )
Quadratwurzel
.RI (\(sr x )
.TP
.BI exp( x )
.RI e^ x
.TP
.BI log( x )
Nat\[:u]rlicher Logarithmus
.TP
.BI log10( x )
Basis-10 Logarithmus
.TP
.BI deg( x )
In Grad umwandeln
.RI ( x " im Bogenma\[ss])"
.TP
.BI rad( x )
Ins Bogenma\[ss] umwandeln
.RI ( x " in Grad)"
.TP
.BI sin( x )
Sinus
.RI ( x " im Bogenma\[ss])"
.TP
.BI sinh( x )
Sinus Hyperbolicus
.RI ( x " im Bogenma\[ss])"
.TP
.BI asin( x )
Arkussinus
.RI ( x " im Bogenma\[ss])"
.TP
.BI cos( x )
Kosinus
.RI ( x " im Bogenma\[ss])"
.TP
.BI cosh( x )
Kosinus Hyperbolicus
.RI ( x " im Bogenma\[ss])"
.TP
.BI acos( x )
Arkuskosinus
.RI ( x " im Bogenma\[ss])"
.TP
.BI tan( x )
Tangens
.RI ( x " im Bogenma\[ss])"
.TP
.BI tanh( x )
Tangens Hyperbolicus
.RI ( x " im Bogenma\[ss])"
.TP
.BI atan( x )
Arkustangens
.RI ( x " im Bogenma\[ss])"
.PP
.BR Achtung :
Der Ausdruck f\[:u]r diese Funktionen muss immer in Klammern stehen!
.PP
Wenn man h\[:a]ufig spezielle Funktionen oder Konstanten ben\[:o]tigt,
kann man sie einfach im Skript hinzuf\[:u]gen.
.PP


NEWS
* AKFAvatar-testing

 - optimized printing of text (line buffered)
 - audio-data with same specs as former one is played with less latency
 - improved support for 24 and 32 bit audio
 - new file Bücher.lua for reading German books from my server (needs curl)

  C-API changes:
    - new macro: AVT_KEY_F

* AKFAvatar 0.24.3

  - audio: big audio files are read directly from file while playing
    (if possible via mmap)
  - audio: AU and WAV data can be read from non-seekable streams
    (loaded into memory)
  - fix: Pascal sound support fixed
  - the control characters SO/SI switch underlined mode on/off;
    the control characters DC1/DC2 switch bold mode on/off
  - configure: new option --with-sdl2
  - SDL2: support for SDL2 with automatic resizing of the graphics
  - SDL2: support for pasting text from clipboard (Alt+V, middle mouse button)

  Lua-API changes:
    - new functions: avt.frequency
    - module "akfavatar-term": fix APC interface

* AKFAvatar 0.24.2

  - lua/clock.lua: resizing works (with some delay)
  - lua/four_in_a_row.lua: mouse support / resizing
  - lua/four_in_a_row.lua: play against computer

  Lua-API changes:
    - Module "akfavatar-graphic":
      - new functions: graphic.set_pointer_buttons_key,
        graphic.set_pointer_motion_key, graphic.get_pointer_position,
        graphic.set_resize_key

* AKFAvatar 0.24.1

  - fix: lua/interactive_lua.lua: fix handling of incomplete lines
  - Lua module "akfavatar-graphic" included in executable
  - text cursor modified
  - lua-akfavatar: GNU .info files can be shown as plain text (no hyperlinks)
  - lua/hexviewer.lua: show more characters on the right side
  - lua/multiplication.lua: clear board (no scrolling)
  - configure accepts the option --with-linuxfb
    which then builds for the Linux framebuffer without the SDL
    (no mouse and no audio support yet)

  C-API changes:
    - new functions: avt_detect_utf8

  Lua-API changes:
    - new functions: avt.toutf8, avt.detect_utf8, avt.utf8codepoints

* AKFAvatar 0.24.0

  - fix: terminal doesn't use so much CPU time anymore
  - fix: window resizing for the pager works again
  - fix: switching fullscreen mode while avatar moves in or out
  - fix: avt_credits sometimes got the end wrong
  - fix: audioplayer.lua accepts PLS playlists again
  - fix: Pascal binding works again, extended
  - paging through menus is animated
  - lua/audioplayer.lua: improved playlist support
  - lua/multiplication.lua: more authentic

  C-API changes:
    - avt_char is unsigned int now, avt_choice adapted
      should be compatible with char32_t in C11
    - functions with *_mb* replaced with functions *_char*
    - new functions: avt_darker, avt_brighter, avt_brightness
    - avtadons.h also uses the prefix avt_ for everything

  Lua-API changes:
    - The default encoding is now ASCII
      to encourage always using avt.encoding()
    - avt.set_directory or avt.chdir return true on success
      or nil and an error message on error

  Pascal-API changes:
    - variable CheckEsc removed
    - new procedures/functions: Menu, ReadKeyCode, reserve_single_keys


* AKFAvatar 0.23.1

  - fixed vulnerabilities
  - avoid recursive calls of the bell
  - split source code for better maintainablitity
  - name sign with 3d effect

  C-API changes:
    - avt_set_title: no automatic charset conversion
    - avt_put_char: support for UTF-16 surrogate characters
    - new functions: avt_wide_copyright, avt_wide_license, avt_wide_version,
      avt_menu, avt_palette_size
    - avtaddons:
      - avta_file_selection and avta_filter have a new parameter "data",
        which is passed along

  Lua-API changes:
    - Module "akfavatar-graphic"
      - new methods: gr:border3d

* AKFAvatar 0.23.0

  The code was made less dependend on SDL. There is a new experimantal
  backend for a Linux framebuffer device, but without mouse support or audio
  (of course you can call an external audio-player in your scripts).

  - bugfix: fix for XPM images with more than 256 colors
  - bugfix: lua-akfavatar also resets the charset
  - bugfix/change: lua-akfavatar option -l sets a global variable
  - avt_update() is much faster
  - support for combining characters (Unicode)
  - The pager supports numbers (number-pad when NumLock is on)
  - lua/interactive_lua.lua supports scrolling back in a history list
  - regression: the navigation bar cannot handle the pause key anymore
  - key codes have been changed not to conflict with the "corporate zone"

  C-API changes:
    - all deprecated functions removed
    - also removed: avt_register_keyhandler, avt_register_mousehandler
    - avt_update() is much faster
    - avt_get_key replaces avt_key
    - modified: avt_show_raw_image, avt_put_raw_image_file,
      avt_put_raw_image_stream, avt_put_raw_image_data, avt_put_raw_image_xpm
      don't have the bpp parameter anymore, only 4 Bytes per pixel accepted
    - new functions: avt_key_pressed, avt_push_key, avt_clear_keys,
      avt_set_audio_end_key, avt_combining, avt_bell_function
    - new experimental functions: avt_set_pointer_buttons_key,
      avt_set_pointer_motion_key, avt_get_pointer_position,
      avt_char avt_input, avt_input_mb
    - new macro: avt_elapsed

  Lua-API changes:
    - the function avt.ask was extended (yet undocumented)
    - new functions: avt.combining, avt.key_pressed, avt.clear_keys,
      avt.push_key, avt.set_audio_end_key, avt.alert, avt.getcwd
    - new table: avt.key with key codes


* AKFAvatar 0.22.1

  There are a lot of changes in the APIs. Also the ABI was changed.

  - Window is resizable again (content gets centered)
  - The image can be a separated banner as header or footer
  - Uses Lua-5.2 now, no longer compatible with 5.1
  - Lua modules don't define globals anymore (scripts have to be adapted!)
  - Data files are put in a separate directory and Lua has a search function
    for it (scripts have to be adapted!)
  - the function for translations is integrated in Lua-AKFAvatar

lua/function_plotter.lua
#!/usr/bin/env lua-akfavatar

--[[-------------------------------------------------------------------
Function Plotter for AKFAvatar
Copyright (c) 2011,2012,2013 Andreas K. Foerster <akf@akfoerster.de>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
--]]-------------------------------------------------------------------

-- how many units to show on the x axis at the default scale
x_units = 2 * 10

-- zoom factor
zoom = 1.25

-- animation when drawing the first time? (true or false)
animation = true

-- allow comma as decimal mark? (true or false)
-- when set to true, you cannot use functions with more than one argument
comma = true

deg = math.deg
rad = math.rad
sqrt = math.sqrt
exp = math.exp
log = math.log
log10 = function(x) return math.log(x, 10); end
pi = math.pi
sin = math.sin
sinh = math.sinh
asin = math.asin
cos = math.cos
cosh = math.cosh
acos = math.acos
tan = math.tan
tanh = math.tanh
atan = math.atan

-- add your own functions and constants here

-------------------------------------------------------------------------------
local avt = require "lua-akfavatar"
local graphic = require "akfavatar-graphic"

local L = avt.translate

avt.translations = {
  ["Function Plotter"] = {
    de="Funktionszeichner" },

  ["Enter the function:"] = {
    de="Gib die Funktion an:" },

  ["The function was not\nin a correct notation!"] = {
    de="Die Funktion war nicht in\nder korrekten Notation!" }
}

avt.encoding("UTF-8")
avt.title(L"Function Plotter")
avt.start()
avt.avatar_image_file(avt.search("teacher.xpm") or "default")

local gr, width, height = graphic.new()
local xoffset, yoffset = width/2, height/2
local scale = width / x_units

local function px(x) -- physical x
  return x * scale + xoffset
end

local function lx(x) -- logical x
  return (x - xoffset) / scale
end

local function py(y) -- physical y
  return y * -scale + yoffset
end

local function ly(y) -- logical y
  return -(y - yoffset) / scale
end

local function drawgrid(step)
  gr:thickness(1)

  if 1 == step then
    gr:color "grey75" -- light for 1's
  elseif 10 == step then
    gr:color "grey55" -- darker for 10's
  else
    return
  end

  for x = step, lx(width), step do
    gr:line (px(x), 1, px(x), height)
  end

  for x = -step, lx(1), -step do
    gr:line (px(x), 1, px(x), height)
  end

  for y = step, ly(1), step do
    gr:line (1, py(y), width, py(y))
  end

  for y = -step, ly(height), -step do
    gr:line (1, py(y), width, py(y))
  end
end

local function grid()
  local s = 5 --> size of the marks
  local step

  gr:thickness(1)
  gr:clear()

  if scale > 10 then
    step = 1
    drawgrid(1)
    drawgrid(10)
  elseif scale > 1 then
    step = 10
    drawgrid(10)
  else
    step = 0
  end

  gr:color "black"

  -- cross
  gr:line (1, yoffset, width, yoffset)
  gr:line (xoffset, 1, xoffset, height)

  -- ticks
  if step > 0 then
    for x = step, lx(width), step do
      gr:line (px(x), yoffset - s,
              px(x), yoffset + s)
    end

    for x = -step, lx(1), -step do
      gr:line (px(x), yoffset - s,
              px(x), yoffset + s)
    end

    for y = step, ly(1), step do
      gr:line (xoffset - s, py(y),
              xoffset + s, py(y))
    end

    for y = -step, ly(height), -step do
      gr:line (xoffset - s, py(y),
              xoffset + s, py(y))
    end
  end
end

local function function_error()
  avt.tell(L"The function was not\nin a correct notation!")
  avt.wait_button()
end

local function reset()
  xoffset = width / 2
  yoffset = height / 2
  scale = width / x_units
end

local function textplate(s)
  local fwidth, fheight = gr:font_size()
  local p = graphic.new((#s*fwidth)+8, fheight+8, "gold")

  p:border3d(false)
  p:center()
  p:textalign("center", "center")
  p:color "black"
  p:text(s)

  return p
end

local function plot(f, fstr)
  local choice
  local animate = animation
  local zoom_factor = zoom
  local functionplate = textplate(fstr)

  reset()

  repeat
    grid()

    if animate then gr:put(functionplate, 4, 4) end

    gr:thickness(2)
    gr:color "royal blue"  --> don't tell my old teacher what pen I use ;-)

    local old_value = f(lx(0))
    gr:moveto(0, py(old_value)) --> offscreen

    for x=1,width do
      local value = f(lx(x))
      local y = py(value)

      -- an invalid number (NaN) is not equal to itself
      if old_value ~= old_value then
        gr:moveto (x, y)
      elseif value*old_value >= 0 or math.abs(old_value - value) < 10
        then gr:lineto (x, y)
        else -- jump with changed sign => assume infinity
          if old_value < 0 then
            gr:lineto (x-1, height)
            gr:line (x, 0, x, y)
          else
            gr:lineto (x-1, 0)
            gr:line (x, height, x, y)
          end
      end

      -- show each time we reach a full value and it's visible
      if animate and x % scale == 0 and y >= 1 and y <= height then
        gr:show()
        avt.wait(0.2)
      end

      old_value = value
    end

    gr:put(functionplate, 4, 4)

    gr:show() --> show final result
    animate = false --> only animate the first time

    choice = avt.navigate "+-udlrsx"
    if "+"==choice then scale = scale * zoom_factor
    elseif "-"==choice then scale = scale / zoom_factor
    elseif "l"==choice then xoffset = xoffset + width/4
    elseif "r"==choice then xoffset = xoffset - width/4
    elseif "u"==choice then yoffset = yoffset + height/4
    elseif "d"==choice then yoffset = yoffset - height/4
    elseif "s"==choice then reset()
    end
  until "x"==choice
end

local function plot_string(s)
  if comma then
    s = string.gsub(s, ",", ".") --> many countries use a comma as decimal mark
  end

  local fs = load("function f(x) return "..s.." end")
  if fs and pcall(fs) then
    -- check if it produces a number
    local okay, v = pcall(f, 1)
    if okay and type(v)=="number" then
      plot(f, "f(x)="..s)
    else
      return function_error()
    end
  else
    return function_error()
  end
end

if arg[1] then
  plot_string(arg[1])
else
  repeat
    avt.set_balloon_size(2, 0)
    avt.say(L"Enter the function:", "\n")
    local funcstring = avt.ask "f(x)="
    if funcstring~="" then plot_string(funcstring) end
  until funcstring == ""
end
